src/main/java/bio/terra/janitor/service/janitor/JanitorService.java

```diff
@@ -2,37 +2,92 @@
 
   @Autowired
-  public JanitorService(JanitorDao janitorDao) {
+  public JanitorService(JanitorDao janitorDao, TransactionTemplate transactionTemplate) {
     this.janitorDao = janitorDao;
+    this.transactionTemplate = transactionTemplate;
   }
 
   public CreatedResource createResource(CreateResourceRequestBody body) {
+    TrackedResourceId id =
+        transactionTemplate.execute(status -> createResourceAndUpdateDuplicates(body, status));
+    return new CreatedResource().id(id.toString());
+  }
+
+  /**
+   * Create a new {@link TrackedResource} and update other resources with the same {@link
+   * CloudResourceUid} as duplicated as appropriate. Returns the created resources {@link
+   * TrackedResourceId}. This should be done as a part of a single database transaction.
+   */
+  private TrackedResourceId createResourceAndUpdateDuplicates(
+      CreateResourceRequestBody body, TransactionStatus unused) {
+    TrackedResourceId id = TrackedResourceId.create(UUID.randomUUID());
     TrackedResource resource =
         TrackedResource.builder()
-            .trackedResourceId(TrackedResourceId.create(UUID.randomUUID()))
+            .trackedResourceId(id)
             .trackedResourceState(TrackedResourceState.READY)
             .cloudResourceUid(body.getResourceUid())
             .creation(body.getCreation().toInstant())
             .expiration(body.getExpiration().toInstant())
             .build();
-
-    // TODO(yonghao): Solution for handling duplicate CloudResourceUid.
+    List<TrackedResource> duplicateResources =
+        janitorDao.retrieveResourcesMatching(
+            TrackedResourceFilter.builder()
+                .cloudResourceUid(resource.cloudResourceUid())
+                .forbiddenStates(
+                    ImmutableSet.of(TrackedResourceState.DONE, TrackedResourceState.DUPLICATED))
+                .build());
+    if (!duplicateResources.isEmpty()) {
+      if (duplicateResources.size() > 1) {
+        // If all resources are created through this function, they should be duplicated
+        // as new resources come in so that there are not multiple resources that need to
+        // be marked as duplicated at once.
+        logger.error(
+            "There was more than one not DONE or DUPLICATE resource with the same CloudResourceUid {}",
+            resource.cloudResourceUid());
+      }
+      Instant lastExpiration =
+          duplicateResources.stream()
+              .map(TrackedResource::expiration)
+              .max(Instant::compareTo)
+              .get();
+      if (resource.expiration().isAfter(lastExpiration)) {
+        // The new resource expires after existing resources. The other resources are now
+        // duplicates of the new resource.
+        for (TrackedResource duplicateResource : duplicateResources) {
+          janitorDao.updateResourceState(
+              duplicateResource.trackedResourceId(), TrackedResourceState.DUPLICATED);
```

Comment:
Do you think ABANDONED should go with the CloudResourceUid or  trackedResourceId?  Shall we mark the new resource as abandon and previous one as DUPLICATED? Or just assume ABANDON is per trackedResourceId thing?

Reply:
I think it should go with the trackedResourceId. Just because something was abandoned once doesn't mean we want any subsequent resources to also be abandoned.

Reply:
Got you. Might want to check with you the abandon endpoint looks like before you left. If we may choose to pull that ticket next week